typed/rosette/base-forms.rkt

```diff
@@ -0,0 +1,390 @@
+   #:with f- (replace-stx-loc #'f-- #'f)
+   ;[⊢ [a ≫ a- ⇒ : τ_a] ...]
+   ;[⊢ [b ≫ b- ⇒ : τ_b] ...]
+   #:with (a- ...) (stx-map expand/ro #'(a ...))
+   #:with (b- ...) (stx-map expand/ro #'(b ...))
+   #:with (τ_a ...) (stx-map typeof #'(a- ...))
+   #:with (τ_b ...) (stx-map typeof #'(b- ...))
+   #:with f* (generate-temporary #'f)
+   #:with (a* ...) (generate-temporaries #'(a ...))
+   #:with (b* ...) (generate-temporaries #'(b ...))
+   #:with [[kw/b* ...] ...] #'[[kw b*] ...]
+   [([f* ≫ _ : τ_f] [a* ≫ _ : τ_a] ... [b* ≫ _ : τ_b] ...)
+    ⊢ [(app f* a* ... kw/b* ... ...) ≫ _ ⇒ : τ_out]]
+   ...
+   #:with [[kw/b- ...] ...] #'[[kw b-] ...]
+   --------
+   [⊢ (ro:#%app f- a- ... kw/b- ... ...) ⇒ (CU τ_out ...)]]
+  ;; symbolic functions
+  [(_ f:expr a:expr ... (~seq kw:keyword b:expr) ...) ≫
+   ;[⊢ [f ≫ f-- ⇒ : (~and (~U* τ_f ...) ~!)]]
+   #:with f-- (expand/ro #'f)
+   #:with (~U* ~! τ_f ...) (typeof #'f--)
+   #:with f- (replace-stx-loc #'f-- #'f)
+   ;[⊢ [a ≫ a- ⇒ : τ_a] ...]
+   ;[⊢ [b ≫ b- ⇒ : τ_b] ...]
+   #:with (a- ...) (stx-map expand/ro #'(a ...))
+   #:with (b- ...) (stx-map expand/ro #'(b ...))
+   #:with (τ_a ...) (stx-map typeof #'(a- ...))
+   #:with (τ_b ...) (stx-map typeof #'(b- ...))
+   #:with f* (generate-temporary #'f)
+   #:with (a* ...) (generate-temporaries #'(a ...))
+   #:with (b* ...) (generate-temporaries #'(b ...))
+   #:with [[kw/b* ...] ...] #'[[kw b*] ...]
+   [([f* ≫ _ : τ_f] [a* ≫ _ : τ_a] ... [b* ≫ _ : τ_b] ...)
+    ⊢ [(app f* a* ... kw/b* ... ...) ≫ _ ⇒ : τ_out]]
+   ...
+   #:with [[kw/b- ...] ...] #'[[kw b-] ...]
+   --------
+   [⊢ (ro:#%app f- a- ... kw/b- ... ...) ⇒ (U τ_out ...)]]
+  ;; symbolic constant functions
+  [(_ f:expr a:expr ... (~seq kw:keyword b:expr) ...) ≫
+   ;[⊢ [f ≫ f-- ⇒ : (~and (~Constant* (~U* τ_f ...)) ~!)]]
+   #:with f-- (expand/ro #'f)
+   #:with (~Constant* (~U* τ_f ...)) (typeof #'f--)
+   #:with f- (replace-stx-loc #'f-- #'f)
+   ;[⊢ [a ≫ a- ⇒ : τ_a] ...]
+   ;[⊢ [b ≫ b- ⇒ : τ_b] ...]
+   #:with (a- ...) (stx-map expand/ro #'(a ...))
+   #:with (b- ...) (stx-map expand/ro #'(b ...))
+   #:with (τ_a ...) (stx-map typeof #'(a- ...))
+   #:with (τ_b ...) (stx-map typeof #'(b- ...))
+   #:with f* (generate-temporary #'f)
+   #:with (a* ...) (generate-temporaries #'(a ...))
+   #:with (b* ...) (generate-temporaries #'(b ...))
+   #:with [[kw/b* ...] ...] #'[[kw b*] ...]
+   [([f* ≫ _ : τ_f] [a* ≫ _ : τ_a] ... [b* ≫ _ : τ_b] ...)
+    ⊢ [(app f* a* ... kw/b* ... ...) ≫ _ ⇒ : τ_out]]
+   ...
+   #:with [[kw/b- ...] ...] #'[[kw b-] ...]
+   --------
+   [⊢ [_ ≫ (ro:#%app f- a- ... kw/b- ... ...) ⇒ : (U τ_out ...)]]])
+
+;; ----------------------------------------------------------------------------
+
+;; Apply
+
+;; This version of apply is very limited: it only works with
+;; functions that declare a rest argument.
+
+(define-typed-syntax apply
+  [(_ f:expr lst:expr) ≫
+   [⊢ f ≫ f- ⇒ (~C→* [] [] #:rest τ_rst τ_out)]
+   [⊢ lst ≫ lst- ⇐ τ_rst]
```

Comment:
No, I believe typed racket made a mistake. Instead I followed the lead of `->*` from the contract library. If we want rest arguments to be able to be heterogeneous (as in `(hash 'a 1 'b 2)` or `(raise-arguments-error 'f "bad" "thing1" thing1 "thing2" thing2)`) then we need to have the list type there.